README.rst
============
pymt_heatcxx
============


.. image:: https://img.shields.io/badge/CSDMS-Basic%20Model%20Interface-green.svg
        :target: https://bmi.readthedocs.io/
        :alt: Basic Model Interface

.. image:: https://img.shields.io/badge/recipe-pymt_heatcxx-green.svg
        :target: https://anaconda.org/conda-forge/pymt_heatcxx

.. image:: https://readthedocs.org/projects/pymt-heatcxx/badge/?version=latest
        :target: https://pymt-heatcxx.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status

.. image:: https://github.com/mdpiper/pymt_heatcxx/actions/workflows/test.yml/badge.svg
        :target: https://github.com/mdpiper/pymt_heatcxx/actions/workflows/test.yml

.. image:: https://github.com/mdpiper/pymt_heatcxx/actions/workflows/flake8.yml/badge.svg
        :target: https://github.com/mdpiper/pymt_heatcxx/actions/workflows/flake8.yml

.. image:: https://github.com/mdpiper/pymt_heatcxx/actions/workflows/black.yml/badge.svg
        :target: https://github.com/mdpiper/pymt_heatcxx/actions/workflows/black.yml


PyMT plugin for heatcxx model


* Free software: MIT License
* Documentation: https://pymt-heatcxx.readthedocs.io.




============ ======================================
Component    PyMT
============ ======================================
HeatModelCxx `from pymt.models import HeatModelCxx`
============ ======================================

---------------
Installing pymt
---------------

Installing `pymt` from the `conda-forge` channel can be achieved by adding
`conda-forge` to your channels with:

.. code::

  conda config --add channels conda-forge

*Note*: Before installing `pymt`, you may want to create a separate environment
into which to install it. This can be done with,

.. code::

  conda create -n pymt python=3
  conda activate pymt

Once the `conda-forge` channel has been enabled, `pymt` can be installed with:

.. code::

  conda install pymt

It is possible to list all of the versions of `pymt` available on your platform with:

.. code::

  conda search pymt --channel conda-forge

-----------------------
Installing pymt_heatcxx
-----------------------



To install `pymt_heatcxx`,

.. code::

  conda install pymt_heatcxx


CHANGES.rst
Changelog for pymt_heatcxx
==========================

0.2 (unreleased)
----------------

- Rename component from HeatModel to HeatModelCxx (#1)
- Show total temperature in BMI example (#2)


0.1 (2021-08-24)
----------------

- Initial release


examples/pymt_heatcxx_ex.py
"""Run the heatcxx model in pymt."""
import numpy as np
from pymt.models import HeatModelCxx

np.set_printoptions(formatter={"float": "{: 6.2f}".format})


# Instantiate the component and get its name.
m = HeatModelCxx()
print(m.name)

# Call setup, then initialize the model.
args = m.setup(".")
m.initialize(*args)

# Get time information from the model.
print("Start time:", m.start_time)
print("End time:", m.end_time)
print("Current time:", m.time)
print("Time step:", m.time_step)
print("Time units:", m.time_units)

# List the model's exchange items.
print("Number of input vars:", len(m.input_var_names))
for var in m.input_var_names:
    print(" - {}".format(var))
print("Number of output vars:", len(m.output_var_names))
for var in m.output_var_names:
    print(" - {}".format(var))

# Get variable info.
var_name = m.output_var_names[0]
print("Variable {}".format(var_name))
print(" - variable type:", m.var_type(var_name))
print(" - units:", m.var_units(var_name))
print(" - itemsize:", m.var_itemsize(var_name))
print(" - nbytes:", m.var_nbytes(var_name))
print(" - location:", m.var_location(var_name))

# Get grid info for variable.
grid_id = m.var_grid(var_name)
print(" - grid id:", grid_id)
print(" - grid type:", m.grid_type(grid_id))
print(" - rank:", m.grid_ndim(grid_id))
grid_size = m.grid_node_count(grid_id)
print(" - size:", grid_size)
print(" - shape:", m.grid_shape(grid_id))

# Get the initial values of the variable.
print("Get initial values of {}...".format(var_name))
print(" - values, flattened:")
print(m.var[var_name].data)
print(" - values, redimensionalized:")
print(m.var[var_name].data.reshape(m.grid_shape(grid_id)))

# Set new values.
print("Set new values of {}...".format(var_name))
new = np.zeros(grid_size, dtype=float)
new[20] = 10.0
m.set_value(var_name, new)
print(" - values, flattened:")
print(m.var[var_name].data)
print(" - values, redimensionalized:")
print(m.var[var_name].data.reshape(m.grid_shape(grid_id)))

# Advance the model by one time step.
m.update()
print("Update: current time:", m.time)
print(" - values at time {}:".format(m.time))
print(m.var[var_name].data.reshape(m.grid_shape(grid_id)))

# Advance the model until a later time.
m.update_until(5.0)
print("Update: current time:", m.time)
print(" - values at time {}:".format(m.time))
print(m.var[var_name].data.reshape(m.grid_shape(grid_id)))

# Finalize the model.
m.finalize()
print("Done.")


examples/heatcxx_ex.py
"""Run the heatcxx model through its BMI in Python."""
import numpy as np
from pymt_heatcxx import HeatModelCxx

config_file = "config.txt"
np.set_printoptions(formatter={"float": "{: 6.2f}".format})


# Instantiate and initialize the model.
m = HeatModelCxx()
print(m.get_component_name())
m.initialize(config_file)

# Get time information from the model.
print("Start time:", m.get_start_time())
print("End time:", m.get_end_time())
print("Current time:", m.get_current_time())
print("Time step:", m.get_time_step())
print("Time units:", m.get_time_units())

# List the model's exchange items.
print("Input vars:", m.get_input_var_names())
print("Output vars:", m.get_output_var_names())

# Get the grid_id for the plate_surface__temperature variable.
var_name = "plate_surface__temperature"
print("Variable {}".format(var_name))
grid_id = m.get_var_grid(var_name)
print(" - grid id:", grid_id)

# Get grid and variable info for plate_surface__temperature.
print(" - grid type:", m.get_grid_type(grid_id))
grid_rank = m.get_grid_rank(grid_id)
print(" - rank:", grid_rank)
grid_shape = np.empty(grid_rank, dtype=np.int32)
m.get_grid_shape(grid_id, grid_shape)
print(" - shape:", grid_shape)
grid_size = m.get_grid_size(grid_id)
print(" - size:", grid_size)
grid_spacing = np.empty(grid_rank, dtype=float)
m.get_grid_spacing(grid_id, grid_spacing)
print(" - spacing:", grid_spacing)
grid_origin = np.empty(grid_rank, dtype=float)
m.get_grid_origin(grid_id, grid_origin)
print(" - origin:", grid_origin)
print(" - variable type:", m.get_var_type(var_name))
print(" - units:", m.get_var_units(var_name))
print(" - itemsize:", m.get_var_itemsize(var_name))
print(" - nbytes:", m.get_var_nbytes(var_name))

# View default initial temperature values.
val = np.empty(grid_size, dtype=float)
m.get_value(var_name, val)
print(" - initial values (flattened):")
print(val)
print(" - initial values (gridded):")
print(val.reshape(grid_shape))
print(" -- total temperature:", val.sum())

# Set new initial temperature values.
new = np.zeros(grid_size, dtype=float)
new[20] = 10.0
m.set_value(var_name, new)
val = np.empty(grid_size, dtype=float)
m.get_value(var_name, val)
print(" - new initial values (flattened):")
print(val)
print(" - new initial values (gridded):")
print(val.reshape(grid_shape))
print(" -- total temperature:", val.sum())

# Advance the model by one time step.
m.update()
print("Updated time:", m.get_current_time())
val = np.empty(grid_size, dtype=float)
m.get_value(var_name, val)
print(" - updated values (gridded):")
print(val.reshape(grid_shape))
print(" -- total temperature:", val.sum())

# # Get a reference to the temperature values and check that it updates.
# print(" - values (by ref, gridded) at time {}:".format(m.get_current_time()))
# ref = m.get_value_ptr(var_name)
# print(ref.reshape(grid_shape))
# m.update()
# print(" - values (by ref, gridded) at time {}:".format(m.get_current_time()))
# print(ref.reshape(grid_shape))

# Advance the model until a later time.
m.update_until(5.0)
print("Later time:", m.get_current_time())
val = np.empty(grid_size, dtype=float)
m.get_value(var_name, val)
print(" - updated values (gridded):")
print(val.reshape(grid_shape))
# print(ref.reshape(grid_shape))
print(" -- total temperature:", val.sum())

# Finalize the model.
m.finalize()
